File: app/resources/mark/slugify_test.go
package mark

import (
	"testing"

	"github.com/stretchr/testify/assert"
)

func TestSlugify(t *testing.T) {
	tests := []struct {
		name     string
		input    string
		expected string
	}{
		{
			name:     "cyrillic",
			input:    "Документация",
			expected: "документация",
		},
		{
			name:     "japanese with spaces",
			input:    "日本語 テスト",
			expected: "日本語-テスト",
		},
		{
			name:     "greek",
			input:    "Παράδειγμα",
			expected: "παράδειγμα",
		},
		{
			name:     "hindi",
			input:    "परीक्षण दस्तावेज़",
			expected: "परीक्षण-दस्तावेज़",
		},
		{
			name:     "korean with full-width space",
			input:    "문서　테스트",
			expected: "문서-테스트",
		},
		{
			name:     "hebrew with hyphen",
			input:    "תיעוד-מערכת",
			expected: "תיעוד-מערכת",
		},
		{
			name:     "persian with hyphen",
			input:    "مثالِ-سادِه",
			expected: "مثالِ-سادِه",
		},
		{
			name:     "basic english",
			input:    "Hello World",
			expected: "hello-world",
		},
		{
			name:     "uppercase to lowercase",
			input:    "HELLO WORLD",
			expected: "hello-world",
		},
		{
			name:     "mixed case",
			input:    "HeLLo WoRLd",
			expected: "hello-world",
		},
		{
			name:     "leading spaces",
			input:    "   hello world",
			expected: "hello-world",
		},
		{
			name:     "trailing spaces",
			input:    "hello world   ",
			expected: "hello-world",
		},
		{
			name:     "leading and trailing spaces",
			input:    "   hello world   ",
			expected: "hello-world",
		},
		{
			name:     "multiple spaces",
			input:    "hello    world",
			expected: "hello-world",
		},
		{
			name:     "multiple hyphens",
			input:    "hello----world",
			expected: "hello-world",
		},
		{
			name:     "leading hyphens",
			input:    "---hello-world",
			expected: "hello-world",
		},
		{
			name:     "trailing hyphens",
			input:    "hello-world---",
			expected: "hello-world",
		},
		{
			name:     "leading underscores",
			input:    "___hello_world",
			expected: "hello_world",
		},
		{
			name:     "trailing underscores",
			input:    "hello_world___",
			expected: "hello_world",
		},
		{
			name:     "special characters",
			input:    "hello@world!test#123",
			expected: "hello-world-test-123",
		},
		{
			name:     "punctuation",
			input:    "hello, world. test?",
			expected: "hello-world-test",
		},
		{
			name:     "brackets and parens",
			input:    "hello (world) [test]",
			expected: "hello-world-test",
		},
		{
			name:     "emojis",
			input:    "hello 👋 world 🌍",
			expected: "hello-world",
		},
		{
			name:     "mixed emojis and text",
			input:    "🎉 Party Time 🎊",
			expected: "party-time",
		},
		{
			name:     "numbers",
			input:    "test 123 456",
			expected: "test-123-456",
		},
		{
			name:     "numbers with letters",
			input:    "test123abc456",
			expected: "test123abc456",
		},
		{
			name:     "accented characters",
			input:    "café résumé",
			expected: "café-résumé",
		},
		{
			name:     "german umlauts",
			input:    "Über Größe",
			expected: "über-größe",
		},
		{
			name:     "full-width characters",
			input:    "ｈｅｌｌｏ　ｗｏｒｌｄ",
			expected: "hello-world",
		},
		{
			name:     "mixed full-width and half-width",
			input:    "hello ｗｏｒｌｄ",
			expected: "hello-world",
		},
		{
			name:     "tabs",
			input:    "hello\tworld",
			expected: "hello-world",
		},
		{
			name:     "newlines",
			input:    "hello\nworld",
			expected: "hello-world",
		},
		{
			name:     "carriage returns",
			input:    "hello\rworld",
			expected: "hello-world",
		},
		{
			name:     "mixed whitespace",
			input:    "hello \t\n\r world",
			expected: "hello-world",
		},
		{
			name:     "zero width space",
			input:    "hello\u200Bworld",
			expected: "hello-world",
		},
		{
			name:     "zero width non-joiner",
			input:    "hello\u200Cworld",
			expected: "hello-world",
		},
		{
			name:     "zero width joiner",
			input:    "hello\u200Dworld",
			expected: "hello-world",
		},
		{
			name:     "soft hyphen",
			input:    "hello\u00ADworld",
			expected: "hello-world",
		},
		{
			name:     "non-breaking space",
			input:    "hello\u00A0world",
			expected: "hello-world",
		},
		{
			name:     "empty string",
			input:    "",
			expected: "",
		},
		{
			name:     "only spaces",
			input:    "     ",
			expected: "",
		},
		{
			name:     "only hyphens",
			input:    "-----",
			expected: "",
		},
		{
			name:     "only special characters",
			input:    "!@#$%^&*()",
			expected: "",
		},
		{
			name:     "only emojis",
			input:    "👋🌍🎉",
			expected: "",
		},
		{
			name:     "url with protocol",
			input:    "https://example.com",
			expected: "https-example-com",
		},
		{
			name:     "email address",
			input:    "user@example.com",
			expected: "user-example-com",
		},
		{
			name:     "path-like string",
			input:    "path/to/file.txt",
			expected: "path-to-file-txt",
		},
		{
			name:     "mixed scripts",
			input:    "English 日本語 Русский",
			expected: "english-日本語-русский",
		},
		{
			name:     "right-to-left scripts",
			input:    "العربية עברית",
			expected: "العربية-עברית",
		},
		{
			name:     "chinese characters",
			input:    "中文测试",
			expected: "中文测试",
		},
		{
			name:     "chinese with spaces",
			input:    "中文 测试",
			expected: "中文-测试",
		},
		{
			name:     "thai",
			input:    "ทดสอบ ภาษาไทย",
			expected: "ทดสอบ-ภาษาไทย",
		},
		{
			name:     "quotes and apostrophes",
			input:    "it's a \"test\"",
			expected: "it-s-a-test",
		},
		{
			name:     "slashes and backslashes",
			input:    "test/slash\\backslash",
			expected: "test-slash-backslash",
		},
		{
			name:     "currency symbols",
			input:    "$100 €50 ¥1000",
			expected: "100-50-1000",
		},
		{
			name:     "math symbols",
			input:    "x + y = z",
			expected: "x-y-z",
		},
		{
			name:     "already valid slug",
			input:    "hello-world",
			expected: "hello-world",
		},
		{
			name:     "underscores preserved",
			input:    "hello_world_test",
			expected: "hello_world_test",
		},
		{
			name:     "mixed hyphens and underscores",
			input:    "hello-world_test",
			expected: "hello-world_test",
		},
		{
			name:     "control characters",
			input:    "hello\x00\x01\x02world",
			expected: "hello-world",
		},
		{
			name:     "ligatures",
			input:    "ﬁle ﬂag",
			expected: "file-flag",
		},
		{
			name:     "superscripts and subscripts",
			input:    "x² + y₃",
			expected: "x2-y3",
		},
		{
			name:     "fractions",
			input:    "½ + ¼",
			expected: "1-2-1-4",
		},
		{
			name:     "combining diacritics",
			input:    "e\u0301" + "cole",
			expected: "école",
		},
		{
			name:     "arabic numerals in arabic",
			input:    "مثال ١٢٣",
			expected: "مثال-١٢٣",
		},
		{
			name:     "devanagari numerals",
			input:    "परीक्षण १२३",
			expected: "परीक्षण-१२३",
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			result := Slugify(tt.input)
			assert.Equal(t, tt.expected, result)
		})
	}
}

File: web/src/utils/slugify.test.ts
import { test } from "uvu";
import * as assert from "uvu/assert";

import { slugify, slugifyDraft } from "./slugify";

test("cyrillic", () => {
  assert.is(slugify("Документация"), "документация");
});

test("japanese with spaces", () => {
  assert.is(slugify("日本語 テスト"), "日本語-テスト");
});

test("greek", () => {
  assert.is(slugify("Παράδειγμα"), "παράδειγμα");
});

test("hindi", () => {
  assert.is(slugify("परीक्षण दस्तावेज़"), "परीक्षण-दस्तावेज़");
});

test("korean with full-width space", () => {
  assert.is(slugify("문서　테스트"), "문서-테스트");
});

test("hebrew with hyphen", () => {
  assert.is(slugify("תיעוד-מערכת"), "תיעוד-מערכת");
});

test("persian with hyphen", () => {
  assert.is(slugify("مثالِ-سادِه"), "مثالِ-سادِه");
});

test("basic english", () => {
  assert.is(slugify("Hello World"), "hello-world");
});

test("uppercase to lowercase", () => {
  assert.is(slugify("HELLO WORLD"), "hello-world");
});

test("mixed case", () => {
  assert.is(slugify("HeLLo WoRLd"), "hello-world");
});

test("leading spaces", () => {
  assert.is(slugify("   hello world"), "hello-world");
});

test("trailing spaces", () => {
  assert.is(slugify("hello world   "), "hello-world");
});

test("leading and trailing spaces", () => {
  assert.is(slugify("   hello world   "), "hello-world");
});

test("multiple spaces", () => {
  assert.is(slugify("hello    world"), "hello-world");
});

test("multiple hyphens", () => {
  assert.is(slugify("hello----world"), "hello-world");
});

test("leading hyphens", () => {
  assert.is(slugify("---hello-world"), "hello-world");
});

test("trailing hyphens", () => {
  assert.is(slugify("hello-world---"), "hello-world");
});

test("leading underscores", () => {
  assert.is(slugify("___hello_world"), "hello_world");
});

test("trailing underscores", () => {
  assert.is(slugify("hello_world___"), "hello_world");
});

test("special characters", () => {
  assert.is(slugify("hello@world!test#123"), "hello-world-test-123");
});

test("punctuation", () => {
  assert.is(slugify("hello, world. test?"), "hello-world-test");
});

test("brackets and parens", () => {
  assert.is(slugify("hello (world) [test]"), "hello-world-test");
});

test("emojis", () => {
  assert.is(slugify("hello 👋 world 🌍"), "hello-world");
});

test("mixed emojis and text", () => {
  assert.is(slugify("🎉 Party Time 🎊"), "party-time");
});

test("numbers", () => {
  assert.is(slugify("test 123 456"), "test-123-456");
});

test("numbers with letters", () => {
  assert.is(slugify("test123abc456"), "test123abc456");
});

test("accented characters", () => {
  assert.is(slugify("café résumé"), "café-résumé");
});

test("german umlauts", () => {
  assert.is(slugify("Über Größe"), "über-größe");
});

test("full-width characters", () => {
  assert.is(slugify("ｈｅｌｌｏ　ｗｏｒｌｄ"), "hello-world");
});

test("mixed full-width and half-width", () => {
  assert.is(slugify("hello ｗｏｒｌｄ"), "hello-world");
});

test("tabs", () => {
  assert.is(slugify("hello\tworld"), "hello-world");
});

test("newlines", () => {
  assert.is(slugify("hello\nworld"), "hello-world");
});

test("carriage returns", () => {
  assert.is(slugify("hello\rworld"), "hello-world");
});

test("mixed whitespace", () => {
  assert.is(slugify("hello \t\n\r world"), "hello-world");
});

test("zero width space", () => {
  assert.is(slugify("hello\u200Bworld"), "hello-world");
});

test("zero width non-joiner", () => {
  assert.is(slugify("hello\u200Cworld"), "hello-world");
});

test("zero width joiner", () => {
  assert.is(slugify("hello\u200Dworld"), "hello-world");
});

test("soft hyphen", () => {
  assert.is(slugify("hello\u00ADworld"), "hello-world");
});

test("non-breaking space", () => {
  assert.is(slugify("hello\u00A0world"), "hello-world");
});

test("empty string", () => {
  assert.is(slugify(""), "");
});

test("only spaces", () => {
  assert.is(slugify("     "), "");
});

test("only hyphens", () => {
  assert.is(slugify("-----"), "");
});

test("only special characters", () => {
  assert.is(slugify("!@#$%^&*()"), "");
});

test("only emojis", () => {
  assert.is(slugify("👋🌍🎉"), "");
});

test("url with protocol", () => {
  assert.is(slugify("https://example.com"), "https-example-com");
});

test("email address", () => {
  assert.is(slugify("user@example.com"), "user-example-com");
});

test("path-like string", () => {
  assert.is(slugify("path/to/file.txt"), "path-to-file-txt");
});

test("mixed scripts", () => {
  assert.is(slugify("English 日本語 Русский"), "english-日本語-русский");
});

test("right-to-left scripts", () => {
  assert.is(slugify("العربية עברית"), "العربية-עברית");
});

test("chinese characters", () => {
  assert.is(slugify("中文测试"), "中文测试");
});

test("chinese with spaces", () => {
  assert.is(slugify("中文 测试"), "中文-测试");
});

test("thai", () => {
  assert.is(slugify("ทดสอบ ภาษาไทย"), "ทดสอบ-ภาษาไทย");
});

test("quotes and apostrophes", () => {
  assert.is(slugify('it\'s a "test"'), "it-s-a-test");
});

test("slashes and backslashes", () => {
  assert.is(slugify("test/slash\\backslash"), "test-slash-backslash");
});

test("currency symbols", () => {
  assert.is(slugify("$100 €50 ¥1000"), "100-50-1000");
});

test("math symbols", () => {
  assert.is(slugify("x + y = z"), "x-y-z");
});

test("already valid slug", () => {
  assert.is(slugify("hello-world"), "hello-world");
});

test("underscores preserved", () => {
  assert.is(slugify("hello_world_test"), "hello_world_test");
});

test("mixed hyphens and underscores", () => {
  assert.is(slugify("hello-world_test"), "hello-world_test");
});

test("control characters", () => {
  assert.is(slugify("hello\x00\x01\x02world"), "hello-world");
});

test("ligatures", () => {
  assert.is(slugify("ﬁle ﬂag"), "file-flag");
});

test("superscripts and subscripts", () => {
  assert.is(slugify("x² + y₃"), "x2-y3");
});

test("fractions", () => {
  assert.is(slugify("½ + ¼"), "1-2-1-4");
});

test("combining diacritics", () => {
  assert.is(slugify("e\u0301cole"), "école");
});

test("arabic numerals in arabic", () => {
  assert.is(slugify("مثال ١٢٣"), "مثال-١٢٣");
});

test("devanagari numerals", () => {
  assert.is(slugify("परीक्षण १२३"), "परीक्षण-१२३");
});

test("draft slug is normalised while typing", () => {
  assert.is(
    slugifyDraft("General!!!!!dnwHDAUIDHAODH3r3u18 2"),
    "general-dnwhdauidhaodh3r3u18-2",
  );
});

test("draft slug preserves one pending word separator", () => {
  assert.is(slugifyDraft("日本語   "), "日本語-");
  assert.is(slugifyDraft("hello___"), "hello_");
});

test("draft slug removes leading separators", () => {
  assert.is(slugifyDraft("---hello"), "hello");
  assert.is(slugifyDraft("---"), "");
});

test.run();
